quixstreams/dataframe/windows/time_based.py

```diff
@@ -127,16 +135,42 @@ def process_window(
             )
         return updated_windows, expired_windows
 
-    def _log_expired_window(self, window, timestamp_ms, late_by_ms) -> None:
+    def _on_expired_window(
+        self,
+        value: Any,
+        key: Any,
+        start: int,
+        end: int,
+        timestamp_ms: int,
+        late_by_ms: int,
+    ) -> None:
         ctx = message_context()
-        logger.warning(
-            "Skipping window processing for expired window "
-            f"timestamp_ms={timestamp_ms} "
-            f"window={window} "
-            f"late_by_ms={late_by_ms} "
-            f"partition={ctx.topic}[{ctx.partition}] "
-            f"offset={ctx.offset}"
-        )
+        to_log = True
+        # Trigger the "on_late" callback if provided.
+        # Log the lateness warning if the callback returns True
+        if on_late := self._on_late:
+            to_log = on_late(
```

Comment:
Double lookup is faster than one lookup and assignment.
```suggestion
        if self._on_late:
            to_log = self._on_late(
```